File: Strings/KnuthMorrisPratt Algorithm.py
# Implementation of KMP Algorithm. Program Author : SNEHA CHAUHAN
''' problem: Given a text txt and a pattern pat, write a function search(char pat[], char txt[]) 
    that prints all occurrences of pat[] in txt[]. 

    KMP algorithm is used to find the pattern in the given string. This is naive approach to find 
    the pattern in the given string.

    Time Complexity: O(n*m)
    Space Complexity: O(1)

    Example: 
    input: txt = “AAAABAAABA” pat = “AAAA”
    output: Pattern found at index 0
        
    Algorithm:
    1. Start from the leftmost character of txt and one by one compare it with each character of pat.
    2. If a character matches, then move both txt and pat ahead and compare the next character.
    3. If a mismatch occurs, then move pat to the index where the mismatch occurs and compare again.
    4. If pat reaches its end without any mismatch, then pattern found.

    '''

def search(pat, txt):
    M = len(pat)
    N = len(txt)

    # A loop to slide pat[] one by one
    for i in range(N - M + 1):
        j = 0

        # For current index i, check for pattern match
        for j in range(0, M):
            if (txt[i + j] != pat[j]):
                break

        if (j == M - 1):
            print("Pattern found at index ", i)


# Driver Code
if __name__ == '__main__':
    txt = "AABAACAADAABAAABAA"
    pat = "AABA"
    search(pat, txt)

File: Strings/KnuthMorrisPrattAlgorithm.cpp
/*
Implementation of KMP Algorithm. 
Program Author : SNEHA CHAUHAN

Problem: Given a text txt and a pattern pat, write a function search(char pat[], char txt[]) 
that prints all occurrences of pat[] in txt[]. 

Appproach:
1. Preprocess pattern to get lps array (longest proper prefix which is also suffix array)
2. Use lps array to traverse the text.
3. Whenever there is a mismatch, move the pointer of pattern based on lps array.
4. If there is a match, move both pointers of text and pattern.

Time Complexity: O(n+m)
Space Complexity: O(n)


*/

#include <bits/stdc++.h>
using namespace std;

void computeLPSArray(char* pat, int M, int* lps);

void KMPSearch(char* pat, char* txt)
{
    int M = strlen(pat);
    int N = strlen(txt);

    int lps[M];

    computeLPSArray(pat, M, lps);

    int i = 0; 
    int j = 0; 
    while (i < N) {
        if (pat[j] == txt[i]) {
            j++;
            i++;
        }

        if (j == M) {
            cout<<"Found pattern at index "<< i - j<<endl;
            j = lps[j - 1];
        }

        else if (i < N && pat[j] != txt[i]) {
            if (j != 0)
                j = lps[j - 1];
            else
                i = i + 1;
        }
    }
}


void computeLPSArray(char* pat, int M, int* lps)
{
    int len = 0;

    lps[0] = 0; 

    int i = 1;
    while (i < M) {
        if (pat[i] == pat[len]) {
            len++;
            lps[i] = len;
            i++;
        }
        else 
        {
            if (len != 0) {
                len = lps[len - 1];
            }
            else 
            {
                lps[i] = 0;
                i++;
            }
        }
    }
}

int main()
{
    char txt[] = "ABABDABACDABABCABAB";
    char pat[] = "ABABCABAB";
    KMPSearch(pat, txt);
    return 0;
}
